storage/api.go

```diff
@@ -32,7 +33,10 @@ type batchItem struct {
 
 // QueryBatch represents a batch of queries to be executed atomically.
 // We use a custom type that mirrors pgx.Batch but also allows introspection for debugging.
-type QueryBatch struct{ items []*batchItem }
+type QueryBatch struct {
+	items []*batchItem
+	mu    sync.Mutex
```

Comment:
I'm in favor of the mutex as it makes so that if we make the same mistake again (and it's happened 3 times now: consensus, runtimes, evm token analyzer), we won't pay the terrible price of a heisenbug.

> not providing the thread-safe version by default also encourages the above.

... Much like how removing seatbelts would encourage more careful driving. It's technically true! Maybe :)
I admit that I haven't been driving the Go-kart (heh) for too long and maybe my perspective would be different with more experience. But I like seatbelts, even when I don't intend to crash.

Sorry for the tone. Go somehow keeps surprising me with its philosophy, not in a good way.

Reply:
Fine with me :+1: